### ISSUE TITLE
Re-add navigationMethods generator flag to generate fetch[FK-Name] methods on generated records

### ISSUE THREAD

**Issue by axelfontaine** [jOOQ/jOOQ#4210]
The navigationMethods flag (as described here: http://www.jooq.org/doc/3.5/manual/code-generation/codegen-records/) is not supported by the Maven plugin and fails with:

Config:

```
<generate>
  <relations>true</relations>
  <records>true</records>
  <navigationMethods>true</navigationMethods>
</generate>
```

`[ERROR] Failed to execute goal org.jooq:jooq-codegen-maven:3.5.4:generate (default) on project boxfuse-console-jooq: Unable to parse configuration of mojo org.jooq:jooq-codegen-maven:3.5.4:generate for parameter navigationMethods: Cannot find 'navigationMethods' in class org.jooq.util.jaxb.Generate -> [Help 1]`

**Comment by lukaseder** [comment]
Thanks for the info. That's a "bug" in the manual. The navigation methods feature was removed in jOOQ 3.0, see http://www.jooq.org/doc/latest/manual/reference/migrating-to-3.0

> ### Feature removals
> Here are some minor features that have been removed in jOOQ 3.0
> - ...
> - The navigation methods and "foreign key setters" are no longer generated in Record classes, as they are useful only to few users and the generated code is very collision-prone.

**Comment by lukaseder** [comment]
Fix will be published with the next manual update

**Comment by axelfontaine** [comment]
Too bad. This seemed like a useful feature.

**Comment by lukaseder** [comment]
I agree, but any implementation that we would provide would heavily depend on assumptions made on naming conventions of client code. Essentially, there are these options:

- Name the methods after the referenced tables (caveat: multiple foreign keys pointing to the same table)
- Name the methods after the foreign key columns (caveat: columns are often suffixed with `_ID`, which seems weird. Besides, there may be composite foreign keys)
- Name the methods after the foreign key constraints (caveat: few people actually name those)

Note that when we removed those methods, we replaced them by more generic API, such as: 

- [`UpdatableRecord.fetchChild(ForeignKey)`](http://www.jooq.org/javadoc/latest/org/jooq/UpdatableRecord.html#fetchChild-org.jooq.ForeignKey-)
- [`UpdatableRecord.fetchChildren(ForeignKey)`](http://www.jooq.org/javadoc/latest/org/jooq/UpdatableRecord.html#fetchChildren-org.jooq.ForeignKey-)
- [`TableRecord.fetchParent(ForeignKey)`](http://www.jooq.org/javadoc/latest/org/jooq/TableRecord.html#fetchParent-org.jooq.ForeignKey-)

**Comment by lukaseder** [comment]
... of course, you're aware of the risk of creating N+1 issues with these methods :)

**Comment by axelfontaine** [comment]
How about a 4th possibility: Name the methods after the foreign key columns and drop any _id suffix if present

As for N+1, yes I can cut myself when you give me a knife, but I don't have to if I know how to use it.

This actually sparked a second thought, dependent on #4211 :

For a table parent and a table child with a parent_id column, the Record & DAO generator could be extended to provide methods like

```Map<ParentRecord, List<ChildRecord>> ChildDAO.fetchParentMap();```

N+1 solved.

**Comment by lukaseder** [comment]
> How about a 4th possibility: Name the methods after the foreign key columns and drop any _id suffix if present

That is extremely clever, and thus: No :-)

I'm a strong proponent of never confusing elegant with clever. If that's really what you want, you can extend the code generator and generate those methods easily. But we will certainly not maintain such API backwards-compatibly. The main reason why we removed this (clever) feature was because we simply didn't think things through enough.

> As for N+1, yes I can cut myself when you give me a knife, but I don't have to if I know how to use it.

I just keep dropping these disclaimers everywhere for people finding issues later on via Google. Nothing personal :-)

> This actually sparked a second thought, dependent on #4211 :

But that would go back to generating methods for navigation, right? The thing we expressly don't want... Or did I miss something?

**Comment by axelfontaine** [comment]
People need to navigate relations in a database. That's a fact.

Either they write all these methods themselves or the code generator relieves them of this burden. Personally I would pick generated navigation methods (even if some may not be needed) over having to write them myself any day.

As for "extremely clever", we may have to agree to disagree as to how using a defacto industry convention should be called ... :-)

**Comment by lukaseder** [comment]
> People need to navigate relations in a database. That's a fact.

I agree with your fact. And I agree with helping people navigate through their database relations. But jOOQ must absolutely stay generic for two reasons:

1. That's what people really want
2. That's what keeps us from supporting these edge cases eternally

> we may have to agree to disagree as to how using a defacto industry convention should be called

Believe me. I've had long discussions before. And there were dozens of ideas, none of them convincing. Most people appreciate the fact that jOOQ makes absolutely no assumptions about their legacy database. We're not going to set a defacto industry convention here.

Now, apart from discussing strategies, what are your specific use-cases? Maybe there's another low-hanging fruit, somewhere?

**Comment by axelfontaine** [comment]
I am afraid my use case falls squarely into the territory covered by the old navigationMethods functionality, which would have saved me writing these methods in my own DAOs (as the generated ones don't play well with Records as discussed in #4211 )

**Comment by lukaseder** [comment]
Can you show some code?

**Comment by axelfontaine** [comment]
Sure.

```
public SubscriptionRecord getSubscription(SubscriptionPeriodRecord subscriptionPeriod) {
    return jooq.fetchOne(SUBSCRIPTION, SUBSCRIPTION.ID.eq(SUBSCRIPTION_PERIOD.SUBSCRIPTION_ID));
}
```

**Comment by lukaseder** [comment]
I think you mean:

```java
public SubscriptionRecord getSubscription(SubscriptionPeriodRecord subscriptionPeriod) {
    return jooq.fetchOne(SUBSCRIPTION, SUBSCRIPTION.ID.eq(subscriptionPeriod.getSubscriptionId()));
}

So, this is not sufficient?

```java
subscriptionPeriod.fetchParent(Keys.FK_SUPE_SUBS); // Example foreign key name
```

(I agree, it reads a bit technical when you have to think about "parent" and "child"... It would be simpler if you could just write `subscriptionPeriod.getSubscription()`)

**Comment by axelfontaine** [comment]
Since the Keys class is already generated, how about using the same source (DB foreign key names) to generate the navigation methods? If the names of the foreign keys need tuning to be prettier, there is a great tool call Flyway to help with the migration ;-)

**Comment by lukaseder** [comment]
It's actually hard to disagree with that. :-)

I'm trying to think of why this wasn't an option back then. I'm trying to construct an edge-case for databases where foreign key names are unique per table (instead of per schema), but I can't find any, because the code generator already disambiguates all those key names when generating the `Keys.java` class... Hmm!

**Comment by lukaseder** [comment]
I'll reopen this issue and think about it again, for jOOQ 3.7. Perhaps, generating `fetchFKName()` methods would fit into the model without producing edge cases...

**Comment by lukaseder** [comment]
By the way, I forgot, but today's jOOQ 3.6.0 will ship with `Result.fetchChildren(ForeignKey)` and `Result.fetchParents(ForeignKey)`: https://github.com/jOOQ/jOOQ/issues/2686

**Comment by kiru** [comment]
Any new information on this issue?

**Comment by lukaseder** [comment]
@kiru: Thank you very much for your ping. No, this is currently not a priority for the next releases.

**Comment by lukaseder** [comment]
I've stumbled upon this issue again, after quite some time. In jOOQ 3.11, we've introduced implicit joins for to-one relationships:
https://blog.jooq.org/2018/02/20/type-safe-implicit-join-through-path-navigation-in-jooq-3-11

This works as follows:

```java
Customer cu = CUSTOMER.as("cu");
 
ctx.select(
      cu.FIRST_NAME,
      cu.LAST_NAME,
      cu.address().city().country().COUNTRY)
   .from(cu)
   .fetch();
```

The generated method names have the following default naming scheme:

- If a child table references the parent table only once, then the parent table's name is used (in this case, this would correspond to `fetchOne[ParentTable]`, for example `CustomerRecord.fetchOneAddress()`
- If a child table references the parent table several times, then the foreign key name is used

In this case, it would also work the same way for the inverse relationship, when fetching child records:

- If a parent table references the child table only once, then the child table's name is used (in this case, this would correspond to `fetch[ChildTable]`, for example `AddressRecord.fetchCustomer()`
- If a parent table references the child table several times, then the foreign key name (on the child table) is used.

Also:

- When navigating to-many relationships, the name is `fetch[ChildTable]`
- When navigating nullable to-one relationships, the name is `fetchOne[ParentTable]`
- When navigating non-nullable to-one relationships, the name is `fetchSingle[ParentTable]`

**Comment by lukaseder** [comment]
See also: https://github.com/jOOQ/jOOQ/issues/5216

**Comment by lukaseder** [comment]
This issue has seen almost no traction by the community.

A far superior way to access parent and child relationships is via nested `ROW` and `MULTISET` expressions. This area has been heavily improved since jOOQ 3.15. Starting with jOOQ 3.17, it will even be possible to use any `Table<R>` expression as a `SelectField<R>` to project nested records, see this article here: https://blog.jooq.org/projecting-type-safe-nested-tablerecords-with-jooq-3-17/

At the same time, all the functionality requested by this issue is present, just in the form of a more generic API, not a generated one. But still, given the previous example:

> ```java
> Customer cu = CUSTOMER.as("cu");
>  
> ctx.select(
>       cu.FIRST_NAME,
>       cu.LAST_NAME,
>       cu.address().city().country().COUNTRY)
>    .from(cu)
>    .fetch();
> ```

Which can now be adapted as follows, starting from jOOQ 3.17, with a single query producing 2 nested records:

```java
Result<Record2<CustomerRecord, CountryRecord>> result =
ctx.select(CUSTOMER, CUSTOMER.address().city().country())
   .from(CUSTOMER)
   .fetch();
```

Compared to this, which runs 4 queries, instead:

```java
for (CustomerRecord customer : ctx.selectFrom(CUSTOMER)) {
    AddressRecord address = customer.fetchAddress();
    CityRecord city = address.fetchCity();
    CountryRecord country = city.fetchCountry();
}
```

It would be much much worse if the traversal was from parent to children, in case of which thousands of queries might ensue.

I simply don't want to encourage people to use this approach. The argument that people "know how to use it occasionally" doesn't count IMO, because for every 1 power user, there are N beginners shooting themselves in the foot (N+1 problem), so the mere fact of having this "convenience" generated will cause poor database interactions.

After all these years, I'm rejecting this feature request of re-adding a feature that was already removed for precisely the above reasons.
